fixed I2S driver for SAI peripheral, and tested on mimxrt1170_evk

DerekSnell opened the pull request:
Update includes several driver fixes.  Tested with Zephyr i2s_speed test application, and added board-specific files and readme to test to replicate the test setup.

test built with command:
`west build -b mimxrt1170_evk_cm7 tests\drivers\i2s\i2s_speed`

fixes #42434
fixes #42436

This branch is based on open PR [Update LPUART async support to fix build failures #44962 ](https://github.com/zephyrproject-rtos/zephyr/pull/44962)

Commit:
drivers: i2s: mcux_sai: add control for purging buffers

stream_disable()'s should not always purge buffers.
And i2s_rx_stream_disable() needs separate control for
purging in_queue and out_queue since app owns buffers
after placed in out_queue for i2s_read()

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: i2s: mcux_sai: allow TX FIFO to drain before disabling

i2s_tx_stream_disable() was disabling SAI transmitter before FIFO
could complete transmission.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: i2s: mcux_sai: fixed i2s_mcux_config()

driver config settings were getting overwritten by APIs that set
default settings, like SAI_GetClassicI2SConfig().  Moved config code
after those APIs.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: i2s: mcux_sai: fixed RX not completing end of stream

RX FIFO watermark setting causing issue where last 16 words received
were stuck in FIFO, and not requesting DMA to move to buffer.  Fixed by
setting watermark to 0.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: i2s: mcux_sai: fixed SAI driver

Includes several driver fixes and improvement to leverage
scatter/gather mode of DMA.  Loads multiple DMA blocks into TCDs.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
tests: drivers: i2s_speed: enable testing on mimxrt1170_evk board

Readme instructions for connecting SAI4 as TX to SAI1 as RX on this
board.  Also includes overlay and Kconfig settings to run this test
on the board.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
drivers: i2s: mcux_sai: fixed driver message pool sizes

Fixed bug if CONFIG_I2S_RX_BLOCK_COUNT and CONFIG_I2S_TX_BLOCK_COUNT
are different sizes.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

hakehuang commented:
@DerekSnell  LGTM, please remember to revert the temp commit 24d5b179707cab7c41a5f58e8f1a62efe947eabd

Commit:
drivers: i2s: mcux_sai: fixed macro name for SAI_HAS_ON_DEMAND_MODE

corrected driver after HAL PR
https://github.com/zephyrproject-rtos/hal_nxp/pull/147

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
tests: drivers: i2s_speed: place mem_slab's in non-cacheable region

If platform supports non-cacheable region, place mem_slab's for I2S
driver buffers there.  This enables drivers using DMA without cache
coherency issues.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

Commit:
boards: mimxrt1170_evk: add additional I2S pinctrl signals

Adds more I2S signals to SAI1 and SAI4, to use with
tests/drivers/i2s/i2s_speed

Signed-off-by: Derek Snell <derek.snell@nxp.com>

NickolasLapp commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -431,72 +533,13 @@ static int i2s_mcux_config(const struct device *dev, enum i2s_dir dir,
-			config.masterSlave = kSAI_Master;
-		}
-	}
+	const bool is_mclk_master = i2s_cfg->options & I2S_OPT_BIT_CLK_MASTER;
```
This line is broken because `I2S_OPT_BIT_CLK_MASTER` is 0, so this is always zero.
https://github.com/zephyrproject-rtos/zephyr/blob/e9b288b7137e5fb6ded18334aaa5a5f9c23ebb17/include/zephyr/drivers/i2s.h#L177

This needs to be a check for `i2s_cfg->options & I2S_OPT_BIT_CLK_SLAVE` or something similar.

NickolasLapp commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -333,13 +434,14 @@ static void enable_mclk_direction(const struct device *dev, bool dir)
 	uint32_t offset = dev_cfg->mclk_pin_offset;
 	uint32_t mask = dev_cfg->mclk_pin_mask;
-	uint32_t *gpr =  (uint32_t *)DT_REG_ADDR(DT_NODELABEL(iomuxcgpr)) + offset;
+	uint32_t *gpr = (uint32_t *)DT_REG_ADDR(DT_NODELABEL(iomuxcgpr)) + offset;
```
This is not producing the correct result. Offset is the raw offset (in bytes), but the order of operations is adding `offset` uint32_t's, so the resultant gpr register is off by 4x.
On the RT1060 this results in register 0x400ac010 being written, when it should be 0x400ac004

Should be: 
```suggestion
	uint32_t *gpr = (uint32_t *)(DT_REG_ADDR(DT_NODELABEL(iomuxcgpr)) + offset);
```

DerekSnell replied:
updated, thanks @NickolasLapp

dleach02 reviewed (approved):
Test passing on my modified board
